SSA/Projects/LLVMRiscV/Pipeline/SelectionDAG.lean

```diff
@@ -0,0 +1,957 @@
+import SSA.Projects.RISCV64.Tactic.SimpRiscV
+import SSA.Projects.LLVMRiscV.Pipeline.mkRewrite
+import SSA.Projects.LLVMRiscV.Pipeline.ConstantMatching
+
+open LLVMRiscV
+
+/- This file implements `DAGCombiner` patterns extracted from the LLVM Risc-V backend.
+  First, we implement the Lean structure that implements the rewrite patterns and then we implement
+  optimizations for LLVM IR and RISC-V.
+  In particular, we implement the patterns supported by LLVM's `GlobalIsel` for RISC-V.
+  Because `GlobalIsel` is hybrid, some of these patterns regard generic IR,
+  while some are target-dependent.
+-/
+
+@[simp_riscv] lemma toType_bv : TyDenote.toType (Ty.riscv (.bv)) = BitVec 64 := rfl
+@[simp_riscv] lemma id_eq1 {α : Type} (x y : α) :  @Eq (Id α) x y = (x = y):= by simp only
+
+structure RISCVPeepholeRewrite (Γ : List Ty) where
+  lhs : Com LLVMPlusRiscV Γ .pure [Ty.riscv .bv]
+  rhs : Com LLVMPlusRiscV Γ .pure [Ty.riscv .bv]
+  correct : lhs.denote = rhs.denote := by simp_lowering <;> bv_decide
+
+def RISCVPeepholeRewriteToRiscvPeephole (self : RISCVPeepholeRewrite Γ) :
+    PeepholeRewrite LLVMPlusRiscV Γ [Ty.riscv .bv] where
+  lhs := self.lhs
+  rhs := self.rhs
+  correct := self.correct
+
+/-! ### visitADD -/
+
+/-
+  (add x, 0) -> x
+-/
+def visitADD_0 : LLVMPeepholeRewriteRefine 64 [Ty.llvm (.bitvec 64)] where
+  lhs := [LV| {
+    ^entry (%x: i64):
+      %c = llvm.mlir.constant (0) : i64
+      %0 = llvm.add  %x, %c : i64
+      llvm.return %0 : i64
+  }]
+  rhs := [LV| {
+    ^entry (%x: i64):
+      llvm.return %x : i64
+  }]
+
+/-
+  add (sext i1 X), 1 -> zext (not i1 X)
+-/
+def visitADD_sameop : LLVMPeepholeRewriteRefine 64 [Ty.llvm (.bitvec 1)] where
+  lhs := [LV| {
+    ^entry (%x: i1):
+      %c = llvm.mlir.constant (1) : i64
+      %0 = llvm.sext %x : i1 to i64
+      %1 = llvm.add %0, %c : i64
+      llvm.return %1 : i64
+  }]
+  rhs := [LV| {
+    ^entry (%x: i1):
+      %0 = llvm.not %x : i1
+      %1 = llvm.zext %0 : i1 to i64
+      llvm.return %1 : i64
+  }]
+
+/-
+  (and x, -1) -> x,
+-/
+def visitADD_minus1 : LLVMPeepholeRewriteRefine 64 [Ty.llvm (.bitvec 64)] where
+  lhs := [LV| {
+    ^entry (%x: i64):
+      %c = llvm.mlir.constant (-1) : i64
+      %0 = llvm.and  %x, %c : i64
+      llvm.return %0 : i64
+  }]
+  rhs := [LV| {
+    ^entry (%x: i64):
+      llvm.return %x : i64
+  }]
+
+/-
+Test the rewrite:
+  fold ((0-A)+B) -> B-A
+-/
+def ZeroMinusAPlusB : LLVMPeepholeRewriteRefine 64 [Ty.llvm (.bitvec 64), Ty.llvm (.bitvec 64)] where
+  lhs := [LV| {
+    ^entry (%a: i64, %b: i64):
+      %0 = llvm.mlir.constant (0) : i64
+      %1 = llvm.sub %0, %a : i64
+      %2 = llvm.add %1, %b : i64
+      llvm.return %2 : i64
+  }]
+  rhs := [LV| {
+    ^entry (%a: i64, %b: i64):
+      %0 = llvm.sub %b, %a : i64
+      llvm.return %0 : i64
+  }]
+
+/-
+Test the rewrite:
+  fold (A+(0-B)) -> A-B
+-/
+def APlusZeroMinusB : LLVMPeepholeRewriteRefine 64 [Ty.llvm (.bitvec 64), Ty.llvm (.bitvec 64)] where
```

Comment:
rather than plus/minus in Lean we use `add` and `sub`, so this sould be `add_zero_sub` or something like that (unless it's called `APlusZeroMinusB` somewhere in LLVM?)